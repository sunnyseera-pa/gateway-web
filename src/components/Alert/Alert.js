/** @jsx jsx */
import { jsx } from '@emotion/react';
import React from 'react';
import PropTypes from 'prop-types';
import Icon from '../Icon';
import { ReactComponent as CloseIcon } from '../../images/icons/close.svg';
import { ReactComponent as CheckIcon } from '../../images/icons/tick.svg';
import { ReactComponent as DangerIcon } from '../../images/icons/danger.svg';
import { ReactComponent as InfoIcon } from '../../images/icons/info.svg';
import { PROP_TYPES_LAYOUTBOX } from '../LayoutBox/LayoutBox.propTypes';

import * as styles from './Alert.styles.js';
import LayoutBox from '../LayoutBox';

const Alert = ({ icon, variant, onClose, children, mt, mb, ml, mr, width, minWidth, maxWidth, dismissable, duration, ...outerProps }) => {
    const [show, setShow] = React.useState(true);

    const handleClose = React.useCallback(() => {
        setShow(false);

        onClose();
    }, []);

    React.useEffect(() => {
        const showTimeout = setTimeout(() => {
            if (!dismissable && show) handleClose();
        }, duration);

        return () => {
            clearTimeout(showTimeout);
        };
    }, [show, dismissable]);

    return (
        show && (
            <LayoutBox {...{ mt, mb, ml, mr, width, minWidth, maxWidth }}>
                <div css={styles.root({ variant })} {...outerProps}>
                    <div css={styles.icon}>
                        {icon}
                        {!icon && variant === 'success' && <Icon svg={<CheckIcon fill='inherit' />} />}
                        {!icon && variant === 'danger' && <Icon svg={<DangerIcon fill='inherit' />} size='lg' />}
                        {!icon && variant === 'warning' && <Icon svg={<DangerIcon fill='inherit' />} size='lg' />}
                        {!icon && variant === 'info' && <Icon svg={<InfoIcon fill='inherit' />} size='lg' />}
                    </div>
                    <div css={styles.content}>{children}</div>
                    {dismissable && <Icon svg={<CloseIcon fill='inherit' />} onClick={handleClose} />}
                </div>
            </LayoutBox>
        )
    );
};

Alert.propTypes = {
    duration: PropTypes.number,
    dismissable: PropTypes.bool,
    onClose: PropTypes.func,
    variant: PropTypes.oneOf(['success', 'warning', 'info', 'danger']).isRequired,
    ...PROP_TYPES_LAYOUTBOX,
};

Alert.defaultProps = {
    duration: 5000,
    dismissable: true,
    className: 'ui-Alert',
    onClose: () => {},
};

export default Alert;
