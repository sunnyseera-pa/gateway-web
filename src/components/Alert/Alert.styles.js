import { css } from '@emotion/react';

export const root =
    ({ variant }) =>
    theme => {
        const {
            font: { size },
            base: { increment },
            colors,
            components: {
                Alert: { variants },
            },
        } = theme;

        return css`
            background: ${colors[variants[variant].background]};
            color: ${colors[variants[variant].color]};
            fill: ${colors[variants[variant].color]};
            border: 2px solid ${colors[variants[variant].borderColor]};
            padding: ${increment * 3}px;
            display: flex;
            border-radius: 4px;
            line-height: 1;
            font-size: ${size.default};
        `;
    };

export const icon = ({ base: { increment } }) => css`
    margin-right: ${increment * 2}px;
`;

export const content = () => css`
    flex-grow: 1;

    > *:last-child {
        margin-bottom: 0;
    }
`;
