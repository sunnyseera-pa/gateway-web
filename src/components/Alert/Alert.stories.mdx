import { useState } from 'react';
import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs';

import Alert from './Alert';

<Meta title='Core/Alert' component={Alert} />

<Canvas>
    <Story name='Variants'>
        {props => {
            return (
                <>
                    {['success', 'info', 'warning', 'danger'].map(variant => (
                        <div>
                            <Alert variant={variant} mb={4}>
                                Lorem ipsum dolor sit amet ad hicio hortus est
                            </Alert>
                        </div>
                    ))}
                </>
            );
        }}
    </Story>
    <Story name='Duration'>
        {props => {
            const [showClosed, setShowClosed] = React.useState(false);
            const handleClose = React.useCallback(() => {
                setShowClosed(true);
            }, []);
            return (
                <div>
                    <Alert variant='success' mb={4} dismissable={false} autoclose onClose={handleClose}>
                        Lorem ipsum dolor sit amet ad hicio hortus est
                    </Alert>
                    {showClosed && 'Alert has been auto-closed!'}
                </div>
            );
        }}
    </Story>
</Canvas>

<ArgsTable of={Alert} />
