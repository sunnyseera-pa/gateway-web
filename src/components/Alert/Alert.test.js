import { render, waitFor } from '@testing-library/react';
import React from 'react';
import Alert from './Alert';

const props = {
    onClose: jest.fn(),
    variant: 'success',
};

let wrapper;

describe('Given the Alert component', () => {
    describe('When it is rendered', () => {
        beforeEach(() => {
            wrapper = render(<Alert {...props}>Content</Alert>, {
                wrapper: Providers,
            });
        });

        it('Then matches the previous snapshot', async () => {
            expect(wrapper.container).toMatchSnapshot();
        });

        describe('And it is dismissable', () => {
            beforeEach(() => {
                wrapper = render(
                    <Alert {...props} dismissable>
                        Content
                    </Alert>,
                    {
                        wrapper: Providers,
                    }
                );
            });

            afterEach(() => {
                jest.clearAllMocks();
            });

            describe('And it is closed', () => {
                beforeEach(() => {
                    const close = wrapper.container.querySelector('[role="button"]');

                    fireEvent.click(close);
                });

                it('Then does not show anything', async () => {
                    await waitFor(() => {
                        expect(wrapper.container.querySelector('.ui-LayoutBox')).toBeNull();
                    });
                });

                it('Then calls onClose', () => {
                    expect(props.onClose).toHaveBeenCalled();
                });
            });
        });

        describe('And autoclose is set', () => {
            beforeEach(() => {
                wrapper = render(
                    <Alert {...props} autoclose autocloseDuration={1000}>
                        Content
                    </Alert>,
                    {
                        wrapper: Providers,
                    }
                );
            });

            afterAll(() => {
                jest.clearAllMocks();
            });

            it('Then does not show anything', async () => {
                await waitFor(() => {
                    expect(wrapper.container.querySelector('.ui-LayoutBox')).toBeNull();
                });
            });

            it('Then calls onClose', () => {
                expect(props.onClose).toHaveBeenCalled();
            });
        });

        describe('And it has an icon specified', () => {
            it('Then does not show other icons', () => {
                wrapper.rerender(
                    <Alert {...props} icon='Icon'>
                        Content
                    </Alert>
                );

                expect(wrapper.container.querySelector('.ui-Alert__icon').innerHTML).toEqual('Icon');
            });
        });
    });
});
