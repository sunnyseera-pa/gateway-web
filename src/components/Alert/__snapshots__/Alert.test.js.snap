// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the Alert component When it is rendered Then matches the previous snapshot 1`] = `
<div>
  <div
    class="ui-LayoutBox css-1k6rz1e"
  >
    <div
      class="ui-Alert css-dtd0fn-Alert"
    >
      <div
        class="ui-Alert__icon css-1x7ex26-Alert"
      >
        <span
          class="ui-Icon css-1k6rz1e css-1nyc6qc-Icon"
        >
          <iconmock
            fill="inherit"
          />
        </span>
      </div>
      <div
        class="css-ungsl1-Alert"
      >
        Content
      </div>
    </div>
  </div>
</div>
`;
